gh-135552: Make the GC clear weakrefs later

nascheme opened the pull request:
Fix two bugs related to the interaction of weakrefs and the garbage
collector. The weakrefs in the `tp_subclasses` dictionary are needed in
order to correctly invalidate type caches (for example, by calling
`PyType_Modified()`).  Clearing weakrefs before calling finalizers causes
the caches to not be correctly invalidated. That can cause crashes since the
caches can refer to invalid objects.  This is fixed by deferring the
clearing of weakrefs to classes and without callbacks until after finalizers
are executed.

The second bug is caused by weakrefs created while running finalizers. Those
weakrefs can be outside the set of unreachable garbage and therefore survive
the `delete_garbage()` phase (where `tp_clear()` is called on objects).
Those weakrefs can expose to Python-level code objects that have had
`tp_clear()` called on them.  See GH-91636 as an example of this kind of
bug. This is fixed be clearing all weakrefs to unreachable objects after
finalizers have been executed.

<!-- gh-issue-number: gh-135552 -->
* Issue: gh-135552
<!-- /gh-issue-number -->
* Issue: gh-132413
* Issue: gh-91636

nascheme committed:
Make the GC clear weakrefs later.

Clear the weakrefs to unreachable objects after finalizers are called.

nascheme committed:
Remove inaccurate comment.

nascheme committed:
Run clear_weakrefs() with world stopped.

nascheme committed:
Defer weakref clears only for refs to classes.

This avoids breaking tests while fixing the issue with tp_subclasses. In
the long term, it would be better to defer the clear of all weakrefs,
not just the ones referring to classes.  However, that is a more
distruptive change and would seem to have a higher chance of breaking
user code.  So, it would not be something to do in a bugfix release.

nascheme committed:
Ensure weakrefs with callbacks are cleared early.

We need to clear those before executing the callback.  Since this
ensures they can't run a second time, we don't need
_PyGC_SET_FINALIZED().  Revise comments.

sergey-miryanov commented on Python/gc.c:
```diff
@@ -921,20 +894,44 @@ handle_weakrefs(PyGC_Head *unreachable, PyGC_Head *old)
+            // weakref clear for all objects, not just types.
+            if (wr->wr_callback != NULL || !PyType_Check(wr->wr_object)) {
+                // _PyWeakref_ClearRef clears the weakref but leaves the
+                // callback pointer intact.  Obscure: it also changes *wrlist.
```
```suggestion
                // callback pointer intact. 
```
nitpick: This is obsolete here.

nascheme replied:
I think that comment is still accurate and useful.  `wrlist` is still changing as a result of calling clear.  We no longer rely on that to progress through the list.  The fact that it changes the list could be suprising and is worth noting, I think.

sergey-miryanov commented on Python/gc.c:
```diff
@@ -921,20 +894,44 @@ handle_weakrefs(PyGC_Head *unreachable, PyGC_Head *old)
-            PyGC_Head *wrasgc;                  /* AS_GC(wr) */
+        PyWeakReference *next_wr;
+        for (PyWeakReference *wr = *wrlist; wr != NULL; wr = next_wr) {
+            next_wr = wr->wr_next;
```
```suggestion
            // Get the next list element to get iterator progress if we omit 
            // clearing of the weakref (because _PyWeakref_ClearRef changes
            // next pointer in the wrlist).
            next_wr = wr->wr_next;
```

nascheme replied:
Okay.

sergey-miryanov commented on Python/gc.c:
```diff
@@ -1751,6 +1802,22 @@ gc_collect_region(PyThreadState *tstate,
+     * *not* be cleared so that caches based on the type version are correctly
+     * invalidated (see GH-135552 as a bug caused by this).
+     */
+    clear_weakrefs(&final_unreachable);
```
Ough! Nice trick with `final_unreachable`!

nascheme committed:
Add NEWS.

nascheme committed:
Add comment about wrlist iteration.

This is a bit trickly because the wrlist can be changing as we
iterate over it.

nascheme committed:
Merge 'origin/main' into gh-135552-wr-clear-later

nascheme committed:
Merge 'origin/main' into gh-135552-wr-clear-later

nascheme committed:
Revise NEWS.

nascheme committed:
Defer clear for weakrefs without callbacks.

This fixes GH-132413 and GH-135552.

nascheme committed:
Add unit test for GH-132413.

sergey-miryanov commented on Python/gc.c:
```diff
@@ -923,23 +894,39 @@ handle_weakrefs(PyGC_Head *unreachable, PyGC_Head *old, bool allow_callbacks)
+                _PyObject_ASSERT((PyObject *)wr, wr->wr_object == Py_None);
             }
 
             if (wr->wr_callback == NULL) {
```
Is it worth to move this check above and remove `if (wr->wr_callback != NULL)`?

nascheme replied:
Yeah that's probably a bit cleaner source code wise.  The compiler should optimize to the same code anyhow though so I'll leave it as it for now.  We still may decide that we only want to clear certain types of weakrefs and then the current structure is better.

neonene commented on Lib/test/test_finalization.py:
```diff
@@ -280,8 +280,9 @@ def test_simple_resurrect(self):
-            self.assertIs(wr(), None)
+            # This used to be None because weakrefs were cleared before
+            # calling finalizers.  Now they are cleared after.
+            self.assertIsNot(wr(), None)
```
The changes to tests look to be taken from the initial commit.  I would like a comment or a test for the weakref with a callback.

nascheme replied:
Yup, that makes sense.

neonene commented on Lib/test/test_finalization.py:
```diff
@@ -388,8 +389,10 @@ def check_resurrecting_chain(self, classes):
             self.assert_survivors(survivor_ids)
-            # XXX desirable?
-            self.assertEqual([wr() for wr in wrs], [None] * N)
+            for wr in wrs:
+                # These values used to be None because weakrefs were cleared
+                # before calling finalizers.  Now they are cleared after.
+                self.assertIsNotNone(wr())
```
The same as above.

neonene commented on Lib/test/test_gc.py:
```diff
@@ -1047,8 +1066,8 @@ class Z:
-        # if called, it means there is a bug in the GC.  The weakref should be
-        # cleared before Z dies.
+        # In older versions of Python, the weakref was cleared by the
+        # gc.  Now it is not cleared and so the callback is run.
```
~The `a.wz` is now cleared by `clear_weakrefs()` rather than by `handle_weakref_callbacks()` in `gc.c`, so the callback will never be called?~  Wait...

nascheme replied:
Yeah, this is another case of the comment coming from an older commit.  I've revert the comment change because the original comment is still accurate.

neonene commented on Lib/test/test_weakref.py:
```diff
@@ -1314,7 +1314,7 @@ def check_len_cycles(self, dict_type, cons):
         # one item may be kept alive inside the iterator
         self.assertIn(n1, (0, 1))
-        self.assertEqual(n2, 0)
+        self.assertIn(n2, (0, 1))
```
The `test_weak_valued_len_cycles()` now passes the original assertion, so the change may not be necessary.

nascheme replied:
Yup, you seem to be correct.  I reverted this change as well.

nascheme committed:
Add additional tests for weakref clearing.

nascheme committed:
Revert unneeded changes to unit tests.

kumaraditya303 commented on Lib/test/test_finalization.py:
```diff
@@ -276,12 +276,17 @@ def test_simple_resurrect(self):
-            self.assertIs(wr(), None)
+            # This used to be None because weakrefs were cleared before
+            # calling finalizers.  Now they are cleared after.
+            self.assertIsNot(wr(), None)
```
```suggestion
            self.assertIsNotNone(wr())
```

nascheme committed:
Use assertIsNone() and assertIsNotNone().

nascheme committed:
Merge 'origin/main' into gh-135552-wr-clear-later

nascheme committed:
Merge 'origin/main' into gh-135552-wr-clear-later

nascheme committed:
Revise InternalDocs/garbage_collector.md.

nascheme committed:
Revise comments about weakref handling.

Consolidate most of the comments related to handling of weakrefs.  This
avoids having the information repeated in different comments.

nascheme committed:
Merge branch 'origin/main' into gh-135552-wr-clear-later
